addons/boa/views.py

```diff
@@ -0,0 +1,112 @@
+from flask import request
+from rest_framework import status as http_status
+
+from addons.base import generic_views
+from addons.boa.models import BoaProvider
+from addons.boa.serializer import BoaSerializer
+from addons.boa.tasks import submit_to_boa
+from api.base.utils import waterbutler_api_url_for
+from framework.auth.decorators import must_be_logged_in
+from framework.celery_tasks.handlers import enqueue_task
+from osf.models import ExternalAccount, AbstractNode
+from osf.utils import permissions
+from website import settings as osf_settings
+from website.project.decorators import must_have_addon, must_have_permission
+
+logger = logging.getLogger(__name__)
+
+
+SHORT_NAME = 'boa'
+FULL_NAME = 'Boa'
+
+boa_account_list = generic_views.account_list(SHORT_NAME, BoaSerializer)
+boa_import_auth = generic_views.import_auth(SHORT_NAME, BoaSerializer)
+boa_deauthorize_node = generic_views.deauthorize_node(SHORT_NAME)
+boa_get_config = generic_views.get_config(SHORT_NAME, BoaSerializer)
+
+
+@must_be_logged_in
+def boa_add_user_account(auth, **kwargs):
+    """Verifies new external account credentials and adds to user's list.
+    This view expects `username` and `password` fields in the JSON body of the request.
+    """
+
+    username = request.json.get('username')
+    password = request.json.get('password')
+    try:
+        boa_client = BoaClient(endpoint=BOA_API_ENDPOINT)
+        boa_client.login(username, password)
+        boa_client.close()
+    except BoaException:
+        return {'message': 'Boa Login failed.'}, http_status.HTTP_401_UNAUTHORIZED
+
+    provider = BoaProvider(account=None, host=BOA_API_ENDPOINT, username=username, password=password)
+    try:
+        provider.account.save()
+    except ValidationError:
+        provider.account = ExternalAccount.objects.get(
+            provider=provider.short_name,
+            provider_id='{}:{}'.format(BOA_API_ENDPOINT, username).lower()
+        )
+        if provider.account.oauth_key != password:
+            provider.account.oauth_key = password
+            provider.account.save()
+    except Exception:
+        return {}
+
+    user = auth.user
+    if not user.external_accounts.filter(id=provider.account.id).exists():
+        user.external_accounts.add(provider.account)
+
+    user.get_or_add_addon('boa', auth=auth)
+    user.save()
+    return {}
+
+
+@must_be_logged_in
+@must_have_addon(SHORT_NAME, 'node')
+@must_have_permission(permissions.WRITE)
+def boa_submit_job(node_addon, **kwargs):
+
+    req_params = request.json
+
+    # Boa addon configuration
+    provider = node_addon.external_account.provider_id
+    parts = provider.rsplit(':', 1)
+    host, username = parts[0], parts[1]
+    password = node_addon.external_account.oauth_key
+
+    # User and project
+    user = kwargs['auth'].user
+    user_guid = user._id
+    project_guid = req_params['data']['nodeId']
+    project_node = AbstractNode.load(project_guid)
```

Comment:
Minor: This could be moved into the `if is_addon_root` conditional to avoid a potentially-unnecessary DB query.

Reply:
fixed in https://github.com/CenterForOpenScience/osf.io/pull/10481/commits/4e09e0e0750524bebfe64cc38569f37d61ac103e